### PR TITLE
2.1 output tags

### LINKED ISSUE #4: 2.1 Output tags

### PR DESCRIPTION
**Pull request by magley**
Closes #4 

You can see the current value of the output tag in `DbManager`.
You can change the current value of the output tag in `DbManager`.
All tags now have a `Value` field, which represents the last measured/sent value between `Server` and some other client app. It's not a 1:1 with the corresponding measure's value (we don't want to worry about them being in sync all the time).
The RTU syncs the valves' `Open` with the appropriate output tag values.
You can no longer click on the RTU checkboxes to open/close the valves.

**Note:**

The tag value is updated like every other piece of data. What will happen to input tags' values if we update e.g. a name? Does it go to 0 (doesn't seem so?).
There's one place in the code where we use a tag to fetch the measure to fetch the measure's value. Could we use the tag's value directly? They may not be consistent.

### CODE REVIEW

File: USca/USca_RTU/MainWindow.xaml.cs [hunk 1]
```diff
@@ -51,7 +57,14 @@ private async void SendData()
                 Thread.Sleep(250);
 
                 await CommService.SendSignalsBatch(_reader.Signals);
+                FetchOutputTagValues();
```
Comment (elkcert) [inline USca/USca_RTU/MainWindow.xaml.cs:60]: Can these two api calls not be concurrent? Also, adding `FetchOutputTagValues` here changes the purpose of the function it's in, so the name `SendData` is no longer accurate.

**Review by elkcert — CHANGES_REQUESTED** [review]
> The tag value is updated like every other piece of data. What will happen to input tags' values if we update e.g. a name? Does it go to 0 (doesn't seem so?).

According to the spec, input tags don't have values. Since we use the same data structure to discretely represent both types of tags we would benefit from validation, to make sure - for example - that a digital tag doesn't have `Unit` set. Although this is not a priority.

>  There's one place in the code where we use a tag to fetch the measure to fetch the measure's value. Could we use the tag's value directly? They may not be consistent.

Same comment as above.

File: USca/USca_RTU/MainWindow.xaml.cs (hunk 1 above)
Comment (magley) [reply USca/USca_RTU/MainWindow.xaml.cs]: [36134cf](https://github.com/magley/untitled-scada-project/pull/21/commits/36134cf7a1a86361fe62fe6d33f938224b7a79fa)

**Comment by magley** [comment]
> > The tag value is updated like every other piece of data. What will happen to input tags' values if we update e.g. a name? Does it go to 0 (doesn't seem so?).
> 
> According to the spec, input tags don't have values. Since we use the same data structure to discretely represent both types of tags we would benefit from validation, to make sure - for example - that a digital tag doesn't have `Unit` set. Although this is not a priority.
> 
> > There's one place in the code where we use a tag to fetch the measure to fetch the measure's value. Could we use the tag's value directly? They may not be consistent.
> 
> Same comment as above.

Any suggestions as to what to do? The main reason I was/am fine with keeping things as-is is this line in `Server::TagService::Update()`:

```C#
db.Tags.Entry(tag).CurrentValues.SetValues(dto);
```

**Comment by elkcert** [comment]
> Any suggestions as to what to do? The main reason I was/am fine with keeping things as-is is this line in Server::TagService::Update():

Maybe modify the setters of the discrete fields so they can't be updated (or default to some value) if the tag's type (or mode) are wrong for the field? Although in any case we shouldn't run into problems even if we don't do this...

**Comment by magley** [comment]
> > Any suggestions as to what to do? The main reason I was/am fine with keeping things as-is is this line in Server::TagService::Update():
> 
> Maybe modify the setters of the discrete fields so they can't be updated (or default to some value) if the tag's type (or mode) are wrong for the field? Although in any case we shouldn't run into problems even if we don't do this...

I've thought about it and tried a few things, but ultimately it would be for the best if we just didn't do anything. It's easier to ignore values of input tags than to prevent them from having any value in the first place.

**Review by elkcert — APPROVED** [review]

**Pull request merged**